### PR TITLE
[MLIR][Conversion] Add new option "use-opencl" for "convert-gpu-to-spirv" conversion pass

### PR DESCRIPTION
**Pull request by silee2**
Current convert-gpu-to-spirv pass has memory space to storage class map hardcoded as spirv::mapMemorySpaceToVulkanStorageClass A new option use-opencl is added to control changing that value to spirv::mapMemorySpaceToOpenCLStorageClass

use-opencl's default value is false and
spirv::mapMemorySpaceToVulkanStorageClass is set as the default memory space to storage class map

when use-opencl is set to true,
spirv::mapMemorySpaceToOpenCLStorageClass is used instead

### CODE REVIEW

**Comment by silee2** [comment]
The PR is one part of breaking down #65539 into smaller ones

File: mlir/test/Conversion/GPUToSPIRV/load-store-opencl.mlir [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+      %12 = arith.addi %arg3, %0 : index
+      // CHECK: %[[INDEX2:.*]] = spirv.IAdd
+      %13 = arith.addi %arg4, %3 : index
+      // CHECK: %[[OFFSET1_0:.*]] = spirv.Constant 0 : i64
+      // CHECK: %[[STRIDE1_1:.*]] = spirv.Constant 4 : i64
+      // CHECK: %[[UPDATE1_1:.*]] = spirv.IMul %[[STRIDE1_1]], %[[INDEX1]] : i64
+      // CHECK: %[[OFFSET1_1:.*]] = spirv.IAdd %[[OFFSET1_0]], %[[UPDATE1_1]] : i64
+      // CHECK: %[[STRIDE1_2:.*]] = spirv.Constant 1 : i64
+      // CHECK: %[[UPDATE1_2:.*]] = spirv.IMul %[[STRIDE1_2]], %[[INDEX2]] : i64
+      // CHECK: %[[OFFSET1_2:.*]] = spirv.IAdd %[[OFFSET1_1]], %[[UPDATE1_2]] : i64
```
Comment (kuhar) [inline mlir/test/Conversion/GPUToSPIRV/load-store-opencl.mlir:65]: Do we care about the exact offset? This check sequence seems verbose and maybe even fragile (assumes exact order of constants and may be broken by unrelated changes in the lowering patterns).

File: mlir/test/Conversion/GPUToSPIRV/load-store-opencl.mlir [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+      // CHECK-NEXT: %[[VAL1:.*]] = spirv.Load "CrossWorkgroup" %[[PTR1]] : f32
+      %14 = memref.load %arg0[%12, %13] : memref<12x4xf32, #spirv.storage_class<CrossWorkgroup>>
+      // CHECK: %[[PTR2:.*]] = spirv.AccessChain %[[ARG1]]{{\[}}{{%.*}}{{\]}}
+      // CHECK-NEXT: %[[VAL2:.*]] = spirv.Load "CrossWorkgroup" %[[PTR2]]
+      %15 = memref.load %arg1[%12, %13] : memref<12x4xf32, #spirv.storage_class<CrossWorkgroup>>
+      // CHECK: %[[VAL3:.*]] = spirv.FAdd %[[VAL1]], %[[VAL2]] : f32
+      %16 = arith.addf %14, %15 : f32
+      // CHECK: %[[PTR3:.*]] = spirv.AccessChain %[[ARG2]]{{\[}}{{%.*}}{{\]}}
+      // CHECK: spirv.Store "CrossWorkgroup" %[[PTR3]], %[[VAL3]] : f32
+      memref.store %16, %arg2[%12, %13] : memref<12x4xf32, #spirv.storage_class<CrossWorkgroup>>
```
Comment (kuhar) [inline mlir/test/Conversion/GPUToSPIRV/load-store-opencl.mlir:76]: Could we split it up into multiple smaller test cases, similar to those in https://github.com/llvm/llvm-project/blob/main/mlir/test/Conversion/GPUToSPIRV/wmma-ops-to-spirv-nv-coop-matrix.mlir?

File: mlir/test/Conversion/GPUToSPIRV/load-store-opencl.mlir (hunk 1 above)
Comment (silee2) [reply mlir/test/Conversion/GPUToSPIRV/load-store-opencl.mlir:65]: @kuhar This is the same example as load-store.mlir in the same folder testing Vulkan lowering. I agree that the code sequence is fragile. @antiagainst What do you think? do you think a separate test for checking code sequence generated for OpenCL kernels is need?  If not, I can replace this test with a smaller test that just checks Crossworkgroup attribute handling.

File: mlir/include/mlir/Conversion/Passes.td [hunk 3]
```diff
@@ -567,7 +567,10 @@ def ConvertGPUToSPIRV : Pass<"convert-gpu-to-spirv", "ModuleOp"> {
   let options = [
     Option<"use64bitIndex", "use-64bit-index",
            "bool", /*default=*/"false",
-           "Use 64-bit integers to convert index types">
+           "Use 64-bit integers to convert index types">,
+    Option<"useOpenCL", "use-opencl",
+           "bool", /*default=*/"false",
+           "Use OpenCL instead of Vulkan">
```
Comment (antiagainst) [inline mlir/include/mlir/Conversion/Passes.td:573]: Maybe make it clear about what it means? Like "Use memory space mappings for OpenCL instead of the default Vulkan".

File: mlir/test/Conversion/GPUToSPIRV/load-store-opencl.mlir (hunk 1 above)
Comment (antiagainst) [reply mlir/test/Conversion/GPUToSPIRV/load-store-opencl.mlir:65]: Yeah would be nice to avoid the excessive check here. Here the goal is more about testing the storage class mapping; the check for indexing and such should be in MemRefToSPIRV. (The Vulkan side was added way early before things are properly structured so you see it's not exactly following that.)

File: mlir/test/Conversion/GPUToSPIRV/load-store-opencl.mlir [hunk 4]
```diff
@@ -0,0 +1,81 @@
+// RUN: mlir-opt -convert-gpu-to-spirv='use-64bit-index=true use-opencl=true' %s -o - | FileCheck %s
+
+module attributes {
+  gpu.container_module,
+  spirv.target_env = #spirv.target_env<
+    #spirv.vce<v1.0, [Kernel, Int64, Addresses], []>, api=OpenCL, #spirv.resource_limits<>>
+} {
+  func.func @load_store(%arg0: memref<12x4xf32, #spirv.storage_class<CrossWorkgroup>>, %arg1: memref<12x4xf32, #spirv.storage_class<CrossWorkgroup>>, %arg2: memref<12x4xf32, #spirv.storage_class<CrossWorkgroup>>) {
```
Comment (antiagainst) [inline mlir/test/Conversion/GPUToSPIRV/load-store-opencl.mlir:8]: Shouldn't this use GPU memory space attributes or numeric numbers as input so we can check the conversion is really going down the OpenCL mapping?

File: mlir/test/Conversion/GPUToSPIRV/load-store-opencl.mlir (hunk 2 above)
Comment (antiagainst) [reply mlir/test/Conversion/GPUToSPIRV/load-store-opencl.mlir:76]: +1 to small dedicated tests.

**Review by antiagainst — CHANGES_REQUESTED** [review]

**Comment by silee2** [comment]
@kuhar @antiagainst Replacing this PR with #69941 since a new pass flag was not required for OpenCL kernel preparation for GPU compilation pipeline.
